In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split 
import time
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
import pickle
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier   
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier

In [2]:
def rfeFeature(indep_X,dep_Y,n):
        rfelist=[]
        
        log_model = LogisticRegression(solver='lbfgs')
        RF = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
       # NB = GaussianNB()
        DT= DecisionTreeClassifier(criterion = 'gini', max_features='sqrt',splitter='best',random_state = 0)
        svc_model = SVC(kernel = 'linear', random_state = 0)
        #knn = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
        rfemodellist=[log_model,svc_model,RF,DT] 
        for i in   rfemodellist:
            print(i)
            log_rfe = RFE(i, n)
            log_fit = log_rfe.fit(indep_X, dep_Y)
            log_rfe_feature=log_fit.transform(indep_X)
            rfelist.append(log_rfe_feature)
        return rfelist

### Explanation of the Code: `rfeFeature`

This function performs Recursive Feature Elimination (RFE) using multiple machine learning models to select the top `n` features from the dataset.

---

### Code Breakdown:

#### **Function Signature**
```python
def rfeFeature(indep_X, dep_Y, n):


**Parameters:**
- **indep_X:** Independent variables (features).
- **dep_Y:** Dependent variable (target labels).
- **n:** Number of top features to select.

**Returns:**
- A list of transformed datasets, where each dataset contains the n most important features selected by a specific model.


### Step 1: Initialize Variables

```python
rfelist = []

- Initializes an empty list to store the transformed datasets with selected features.


### Step 2: Define Models for RFE


```python
log_model = LogisticRegression(solver='lbfgs')
RF = RandomForestClassifier(n_estimators=10, criterion='entropy', random_state=0)
DT = DecisionTreeClassifier(criterion='gini', max_features='sqrt', splitter='best', random_state=0)
svc_model = SVC(kernel='linear', random_state=0)

- Defines several machine learning models for performing RFE:
  - **log_model:** Logistic Regression.
  - **RF:** Random Forest Classifier.
  - **DT:** Decision Tree Classifier.
  - **svc_model:** Support Vector Classifier (SVC) with a linear kernel.


### Step 3: Model List

```python
rfemodellist = [log_model, svc_model, RF, DT]

- A list of models used to perform RFE.


### Step 4: Perform RFE for Each Model

```python
for i in rfemodellist:
    print(i)
    log_rfe = RFE(i, n)
    log_fit = log_rfe.fit(indep_X, dep_Y)
    log_rfe_feature = log_fit.transform(indep_X)
    rfelist.append(log_rfe_feature)

- Loops through each model in the `rfemodellist`:
  - Prints the current model.
  - Creates an RFE object for the model with `n` features.
  - Fits the RFE object on the dataset (`indep_X` and `dep_Y`).
  - Transforms the dataset to include only the selected `n` features.
  - Appends the transformed dataset to `rfelist`.


### Step 5: Return the List

```python
return rfelist

- Returns the list of datasets (`rfelist`), where each dataset corresponds to the top `n` features selected by a specific model.


**Purpose:**
- **Feature Selection:** Reduces the dimensionality of the dataset by selecting only the most important features.
- **Model-Specific Selection:** Allows different models to guide the feature selection process based on their unique perspectives.


In [3]:
def split_scalar(indep_X,dep_Y):
        X_train, X_test, y_train, y_test = train_test_split(indep_X, dep_Y, test_size = 0.25, random_state = 0)
        #X_train, X_test, y_train, y_test = train_test_split(indep_X,dep_Y, test_size = 0.25, random_state = 0)
        
        #Feature Scaling
        #from sklearn.preprocessing import StandardScaler
        sc = StandardScaler()
        X_train = sc.fit_transform(X_train)
        X_test = sc.transform(X_test)
        
        return X_train, X_test, y_train, y_test

In [4]:
def cm_prediction(classifier,X_test):
     y_pred = classifier.predict(X_test)
        
        # Making the Confusion Matrix
     from sklearn.metrics import confusion_matrix
     cm = confusion_matrix(y_test, y_pred)
        
     from sklearn.metrics import accuracy_score 
     from sklearn.metrics import classification_report 
        #from sklearn.metrics import confusion_matrix
        #cm = confusion_matrix(y_test, y_pred)
        
     Accuracy=accuracy_score(y_test, y_pred )
        
     report=classification_report(y_test, y_pred)
     return  classifier,Accuracy,report,X_test,y_test,cm

In [5]:
def logistic(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.linear_model import LogisticRegression
        classifier = LogisticRegression(random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm      

In [6]:
def svm_linear(X_train,y_train,X_test):
                
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'linear', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [7]:
def svm_NL(X_train,y_train,X_test):
                
        from sklearn.svm import SVC
        classifier = SVC(kernel = 'rbf', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [8]:
def Navie(X_train,y_train,X_test):       
        # Fitting K-NN to the Training set
        from sklearn.naive_bayes import GaussianNB
        classifier = GaussianNB()
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm     

In [9]:
def knn(X_train,y_train,X_test):
           
        # Fitting K-NN to the Training set
        from sklearn.neighbors import KNeighborsClassifier
        classifier = KNeighborsClassifier(n_neighbors = 5, metric = 'minkowski', p = 2)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [10]:
def Decision(X_train,y_train,X_test):
        
        # Fitting K-NN to the Training set
        from sklearn.tree import DecisionTreeClassifier
        classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm     

In [11]:
def random(X_train,y_train,X_test):
        
        # Fitting K-NN to the Training set
        from sklearn.ensemble import RandomForestClassifier
        classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
        classifier.fit(X_train, y_train)
        classifier,Accuracy,report,X_test,y_test,cm=cm_prediction(classifier,X_test)
        return  classifier,Accuracy,report,X_test,y_test,cm

In [12]:
def rfe_classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf): 
    
    rfedataframe=pd.DataFrame(index=['Logistic','SVC','Random','DecisionTree'],columns=['Logistic','SVMl','SVMnl',
                                                                                        'KNN','Navie','Decision','Random'])

    for number,idex in enumerate(rfedataframe.index):
        
        rfedataframe['Logistic'][idex]=acclog[number]       
        rfedataframe['SVMl'][idex]=accsvml[number]
        rfedataframe['SVMnl'][idex]=accsvmnl[number]
        rfedataframe['KNN'][idex]=accknn[number]
        rfedataframe['Navie'][idex]=accnav[number]
        rfedataframe['Decision'][idex]=accdes[number]
        rfedataframe['Random'][idex]=accrf[number]
    return rfedataframe

In [14]:
dataset1=pd.read_csv("prep.csv",index_col=None)
df2=dataset1
df2 = pd.get_dummies(df2, dtype=int, drop_first=True)

indep_X=df2.drop('classification_yes', 1)
dep_Y=df2['classification_yes']

TypeError: DataFrame.drop() takes from 1 to 2 positional arguments but 3 were given

The error occurs because the DataFrame.drop() method in pandas no longer accepts 1 or 0 as positional arguments for the axis parameter in modern versions of pandas. Instead, you should explicitly specify axis=1 for dropping columns or axis=0 for dropping rows.

Here's the corrected version of the above code:

In [15]:
# Load the dataset
dataset1 = pd.read_csv("prep.csv", index_col=None)

# Create a copy of the dataset
df2 = dataset1

# Perform one-hot encoding with drop_first=True
df2 = pd.get_dummies(df2, dtype = int,drop_first=True)

# Separate independent variables (X) and dependent variable (Y)
indep_X = df2.drop('classification_yes', axis=1)  # Specify axis=1 explicitly
dep_Y = df2['classification_yes']

# Check the results
print(indep_X.head())
print(dep_Y.head())


   age         bp   al   su         bgr         bu        sc         sod  \
0  2.0  76.459948  3.0  0.0  148.112676  57.482105  3.077356  137.528754   
1  3.0  76.459948  2.0  0.0  148.112676  22.000000  0.700000  137.528754   
2  4.0  76.459948  1.0  0.0   99.000000  23.000000  0.600000  138.000000   
3  5.0  76.459948  1.0  0.0  148.112676  16.000000  0.700000  138.000000   
4  5.0  50.000000  0.0  0.0  148.112676  25.000000  0.600000  137.528754   

        pot       hrmo  ...  rbc_normal  pc_normal  pcc_present  ba_present  \
0  4.627244  12.518156  ...           1          0            0           0   
1  4.627244  10.700000  ...           1          1            0           0   
2  4.400000  12.000000  ...           1          1            0           0   
3  3.200000   8.100000  ...           1          1            0           0   
4  4.627244  11.800000  ...           1          1            0           0   

   htn_yes  dm_yes  cad_yes  appet_yes  pe_yes  ane_yes  
0        0

In [16]:
rfelist=rfeFeature(indep_X,dep_Y,3)       

acclog=[]
accsvml=[]
accsvmnl=[]
accknn=[]
accnav=[]
accdes=[]
accrf=[]

LogisticRegression()


TypeError: RFE.__init__() takes 2 positional arguments but 3 were given

The error occurs because the RFE constructor is being incorrectly called with three arguments, while it only expects two positional arguments: the estimator (model) and the number of features to select.

**To fix this error, ensure the RFE object is instantiated correctly. Here's the corrected code:**


In [17]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC

def rfeFeature(indep_X, dep_Y, n):
    rfelist = []
    
    # Define models
    log_model = LogisticRegression(solver='lbfgs', max_iter=1000)
    RF = RandomForestClassifier(n_estimators=10, criterion='entropy', random_state=0)
    DT = DecisionTreeClassifier(criterion='gini', max_features='sqrt', splitter='best', random_state=0)
    svc_model = SVC(kernel='linear', random_state=0)
    
    # List of models to use in RFE
    rfemodellist = [log_model, svc_model, RF, DT]
    
    for i in rfemodellist:
        print(i)
        # Correct RFE instantiation
        log_rfe = RFE(estimator=i, n_features_to_select=n)
        log_fit = log_rfe.fit(indep_X, dep_Y)
        log_rfe_feature = log_fit.transform(indep_X)
        rfelist.append(log_rfe_feature)
    
    return rfelist


**Explanation of the Fix:**
- The correct arguments for RFE are:
  - **estimator:** The model (e.g., LogisticRegression, RandomForestClassifier).
  - **n_features_to_select:** The number of features to select (`n` in your function).


**How to use:**

`rfelist = rfeFeature(indep_X, dep_Y, 3)`

This will now execute without errors, and the `rfelist` variable will contain the top 3 selected features for each model in the `rfemodellist`.


In [18]:
for i in rfelist:   
    X_train, X_test, y_train, y_test=split_scalar(i,dep_Y)   
    
        
    classifier,Accuracy,report,X_test,y_test,cm=logistic(X_train,y_train,X_test)
    acclog.append(Accuracy)
    
    classifier,Accuracy,report,X_test,y_test,cm=svm_linear(X_train,y_train,X_test)  
    accsvml.append(Accuracy)
    
    classifier,Accuracy,report,X_test,y_test,cm=svm_NL(X_train,y_train,X_test)  
    accsvmnl.append(Accuracy)
    
    classifier,Accuracy,report,X_test,y_test,cm=knn(X_train,y_train,X_test)  
    accknn.append(Accuracy)
    
    classifier,Accuracy,report,X_test,y_test,cm=Navie(X_train,y_train,X_test)  
    accnav.append(Accuracy)
    
    classifier,Accuracy,report,X_test,y_test,cm=Decision(X_train,y_train,X_test)  
    accdes.append(Accuracy)
    
    classifier,Accuracy,report,X_test,y_test,cm=random(X_train,y_train,X_test)  
    accrf.append(Accuracy)
    
result=rfe_classification(acclog,accsvml,accsvmnl,accknn,accnav,accdes,accrf)

NameError: name 'rfelist' is not defined

The error `NameError: name 'rfelist' is not defined` occurs because the variable `rfelist` has not been defined or assigned a value prior to the loop. You need to ensure that `rfelist` is properly initialized before attempting to iterate over it.

**Steps to Fix:**
- Make sure the `rfeFeature` function is correctly defined and executed to create the `rfelist`.
- Call the `rfeFeature` function with appropriate inputs before the loop.


**Here’s the corrected workflow:**


In [19]:
# Ensure rfeFeature is correctly defined
rfelist = rfeFeature(indep_X, dep_Y, 3)  # Generate rfelist with top 3 features

# Lists to store accuracies for each model
acclog = []
accsvml = []
accsvmnl = []
accknn = []
accnav = []
accdes = []
accrf = []

# Iterating over rfelist and evaluating models
for i in rfelist:   
    X_train, X_test, y_train, y_test = split_scalar(i, dep_Y)  # Ensure split_scalar is defined
    
    # Logistic Regression
    classifier, Accuracy, report, X_test, y_test, cm = logistic(X_train, y_train, X_test)
    acclog.append(Accuracy)
    
    # SVM (Linear Kernel)
    classifier, Accuracy, report, X_test, y_test, cm = svm_linear(X_train, y_train, X_test)  
    accsvml.append(Accuracy)
    
    # SVM (Non-Linear Kernel)
    classifier, Accuracy, report, X_test, y_test, cm = svm_NL(X_train, y_train, X_test)  
    accsvmnl.append(Accuracy)
    
    # k-NN
    classifier, Accuracy, report, X_test, y_test, cm = knn(X_train, y_train, X_test)  
    accknn.append(Accuracy)
    
    # Naive Bayes
    classifier, Accuracy, report, X_test, y_test, cm = Navie(X_train, y_train, X_test)  
    accnav.append(Accuracy)
    
    # Decision Tree
    classifier, Accuracy, report, X_test, y_test, cm = Decision(X_train, y_train, X_test)  
    accdes.append(Accuracy)
    
    # Random Forest
    classifier, Accuracy, report, X_test, y_test, cm = random(X_train, y_train, X_test)  
    accrf.append(Accuracy)

# Generate the result DataFrame
result = rfe_classification(acclog, accsvml, accsvmnl, accknn, accnav, accdes, accrf)


LogisticRegression(max_iter=1000)


C:\Users\dell\anaconda3\envs\aiml\lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


SVC(kernel='linear', random_state=0)
RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)
DecisionTreeClassifier(max_features='sqrt', random_state=0)


C:\Users\dell\AppData\Local\Temp\ipykernel_2044\2386145136.py:8: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  rfedataframe['Logistic'][idex]=acclog[number]
C:\Users\dell\AppData\Local\Temp\ipykernel_2044\2386145136.py:9: FutureWarning: Chai

**Key Points:**
- **Define rfelist:** Call `rfeFeature` before the loop to generate `rfelist`. Ensure `indep_X`, `dep_Y`, and `n` are correctly passed.
- **Define Functions:** Ensure all functions used (`split_scalar`, `logistic`, `svm_linear`, `svm_NL`, `knn`, `Navie`, `Decision`, `random`, `rfe_classification`) are properly implemented and imported before running the code.
- **Variable Scope:** Confirm all variables (`indep_X`, `dep_Y`) are defined in the same scope as this code.

This setup ensures `rfelist` is properly defined, and the loop executes without errors.


To generate a result similar to the output from `SelectKBest` (i.e., a table comparing model accuracies for a given feature selection method), we need to ensure the following steps are taken:

### Generate Consistent Outputs:

- Ensure that the `rfeFeature` function returns the selected features correctly.
- Ensure all models are evaluated in the same way as they were in the `SelectKBest` example.

### Store and Compare Results:

- After obtaining accuracies for each model using `rfeFeature`, organize them into a DataFrame.
- Compare the DataFrame from `SelectKBest` with the one generated by `rfeFeature`.

### Here’s the complete workflow:

#### Step 1: Ensure `rfeFeature` Works Correctly
Make sure `rfeFeature` outputs the reduced feature set correctly for all iterations. For example:


In [20]:
rfelist = rfeFeature(indep_X, dep_Y, 3)  # Generate top 3 features using RFE


LogisticRegression(max_iter=1000)


C:\Users\dell\anaconda3\envs\aiml\lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


SVC(kernel='linear', random_state=0)
RandomForestClassifier(criterion='entropy', n_estimators=10, random_state=0)
DecisionTreeClassifier(max_features='sqrt', random_state=0)


### Step 2: Evaluate Models and Collect Accuracies
Use a similar loop to evaluate models as in the `SelectKBest` example:


In [21]:
# Lists to store accuracies for RFE
acclog_rfe = []
accsvml_rfe = []
accsvmnl_rfe = []
accknn_rfe = []
accnav_rfe = []
accdes_rfe = []
accrf_rfe = []

# Iterate through RFE-transformed datasets
for i in rfelist:
    X_train, X_test, y_train, y_test = split_scalar(i, dep_Y)  # Ensure this works properly

    # Logistic Regression
    classifier, Accuracy, report, X_test, y_test, cm = logistic(X_train, y_train, X_test)
    acclog_rfe.append(Accuracy)
    
    # SVM (Linear Kernel)
    classifier, Accuracy, report, X_test, y_test, cm = svm_linear(X_train, y_train, X_test)
    accsvml_rfe.append(Accuracy)
    
    # SVM (Non-Linear Kernel)
    classifier, Accuracy, report, X_test, y_test, cm = svm_NL(X_train, y_train, X_test)
    accsvmnl_rfe.append(Accuracy)
    
    # k-NN
    classifier, Accuracy, report, X_test, y_test, cm = knn(X_train, y_train, X_test)
    accknn_rfe.append(Accuracy)
    
    # Naive Bayes
    classifier, Accuracy, report, X_test, y_test, cm = Navie(X_train, y_train, X_test)
    accnav_rfe.append(Accuracy)
    
    # Decision Tree
    classifier, Accuracy, report, X_test, y_test, cm = Decision(X_train, y_train, X_test)
    accdes_rfe.append(Accuracy)
    
    # Random Forest
    classifier, Accuracy, report, X_test, y_test, cm = random(X_train, y_train, X_test)
    accrf_rfe.append(Accuracy)


### Step 3: Create DataFrame for RFE Results
Organize the accuracies into a DataFrame:


In [22]:
import pandas as pd

# Create DataFrame for RFE results
rfe_results = pd.DataFrame({
    'Logistic': acclog_rfe,
    'SVMl': accsvml_rfe,
    'SVMnl': accsvmnl_rfe,
    'KNN': accknn_rfe,
    'Navie': accnav_rfe,
    'Decision': accdes_rfe,
    'Random': accrf_rfe
}, index=['RFE'])

print("RFE Results:")
print(rfe_results)


ValueError: Length of values (4) does not match length of index (1)

The error `ValueError: Length of values (4) does not match length of index (1)` indicates that the lists (`acclog_rfe`, `accsvml_rfe`, etc.) contain multiple values (e.g., 4 values) while you have specified only one row label (`index=['RFE']`) for the DataFrame.

This mismatch happens because the lists likely represent multiple iterations (e.g., different feature subsets from RFE), but the DataFrame is expecting only one row.

**Solution**
Modify the code to ensure the DataFrame index matches the number of rows in the lists. Here’s how you can resolve it:

### Step 1: Check List Lengths
Before creating the DataFrame, confirm the length of each list:


In [23]:
print(len(acclog_rfe))  # Check the length of any accuracy list


4


Since the length of `acclog_rfe` is 4, it means you have results for 4 different feature subsets from the RFE process. To proceed, you need to ensure the index of the DataFrame has 4 labels, corresponding to these subsets.

### Here’s how to proceed:

#### Step 1: Create Index Labels
Generate labels for each feature subset:


In [24]:
index_labels = [f'RFE_Set_{i+1}' for i in range(len(acclog_rfe))]


#### This will create:

In [25]:
['RFE_Set_1', 'RFE_Set_2', 'RFE_Set_3', 'RFE_Set_4']


['RFE_Set_1', 'RFE_Set_2', 'RFE_Set_3', 'RFE_Set_4']

### Step 2: Create the DataFrame
Now use these labels to create the `rfe_results` DataFrame:


In [26]:
import pandas as pd

# Create DataFrame for RFE results
rfe_results = pd.DataFrame({
    'Logistic': acclog_rfe,
    'SVMl': accsvml_rfe,
    'SVMnl': accsvmnl_rfe,
    'KNN': accknn_rfe,
    'Navie': accnav_rfe,
    'Decision': accdes_rfe,
    'Random': accrf_rfe
}, index=index_labels)

print("RFE Results:")
print(rfe_results)


RFE Results:
           Logistic  SVMl  SVMnl   KNN  Navie  Decision  Random
RFE_Set_1      0.94  0.94   0.94  0.94   0.94      0.94    0.94
RFE_Set_2      0.87  0.87   0.87  0.87   0.87      0.87    0.87
RFE_Set_3      0.94  0.94   0.94  0.94   0.90      0.91    0.92
RFE_Set_4      0.98  0.98   0.98  0.98   0.79      0.97    0.97


|         | Logistic | SVMl  | SVMnl | KNN   | Navie | Decision | Random |
|---------|----------|-------|-------|-------|-------|----------|--------|
| RFE_Set_1 | 0.94     | 0.94  | 0.94  | 0.94  | 0.94  | 0.94     | 0.94   |
| RFE_Set_2 | 0.87     | 0.87  | 0.87  | 0.87  | 0.87  | 0.87     | 0.87   |
| RFE_Set_3 | 0.94     | 0.94  | 0.94  | 0.94  | 0.90  | 0.91     | 0.92   |
| RFE_Set_4 | 0.98     | 0.98  | 0.98  | 0.98  | 0.79  | 0.97     | 0.97   |
